### DISCUSSION TITLE
how to lock uaysnc server while a task is in use

### DISCUSSION THREAD

**Discussion by GM-Script-Writer-62850** [micropython/micropython#11940] [General]
I have a W5500 PCIO controller, and i have core 1 doing things at time, if i have activity on core 1 when i get input on the server things go south without any error, when i have core 1 active i have a `uasyncio.Lock()`, how can i have this respect the lock (when do i acquire/release it?)
```
async def webAPI(reader, writer):
	request_line = await reader.readline()
	print("\nClient connected:",request_line)
	if(request_line == b""):
		writer.write('HTTP/1.0 204 No Content')# is this right?
		await writer.drain()
		await writer.wait_closed()
		return print("Disconnect")

	length=0
	header=await reader.readline()
	while header != b"\r\n":
		if(header != b""):
			line=header.decode("utf-8")
			if line[0:15] == "Content-Length:":
				length=int(line[15:])
				print("Content length:",length)
			else:
				pass#print('Ignore header:',line)
		header=await reader.readline()

	request=request_line.decode("utf-8").split(' ')
	if request[0]=="POST":
		data=await reader.readexactly(length)
		data=data.decode("utf-8")
		data=json_decode(data)
		print(data)
		if data["action"] == "buzzer":
			uasyncio.create_task(buzzer(data["value"]))
		elif data["action"] == "blink":
			if not data["value"]:
				notifyControlActive=0
			elif not notifyControlActive:
				notifyControlActive=1
				uasyncio.create_task(notifyControl())
		writer.write('HTTP/1.0 200 OK\r\nContent-type: text/plain\r\nContent-Length: 2\r\nCache-Control: no-cache\r\n\r\nOK')
	else:
		writer.write("")
	try:
		await writer.drain()
		await writer.wait_closed()
		print("Client disconnected")
	except Exception as e:
		print("Error:",e)
uasyncio.create_task(uasyncio.start_server(webAPI, "0.0.0.0", 80))
```

**Comment by peterhinch** [comment]
`asyncio.Lock` is not threadsafe. You need to use `_thread.allocate_lock` for a threasafe lock. You may want to read [this doc](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/THREADING.md) which lists some of the issues linking `asyncio` code and threaded code.

**Reply by GM-Script-Writer-62850** [reply]
Lets say i am at `request_line = await reader.readline()` on core 0 (1st line in the function)

now lets say core 1 runs urequest, i need to prevent core 0 from using the network, it can do anything else, just not that

if both core try to use the w5500 at the same time i end up with a lockup, i want to prevent that

**Reply by bixb922** [reply]
Best do all i/o on core 1.

**Reply by GM-Script-Writer-62850** [reply]
but  uasyncio can only run on a single core...

the uasyncio version of urequest does not support sending data with post

**Reply by bixb922** [reply]
Yes, that's true, and asyncio can run many, many tasks on that single core 1. It's not easy to tame that second core, so personally I'd use the second core only when running out of cpu on the first core, or for example for long running cpu intensive tasks.

**Reply by GM-Script-Writer-62850** [reply]
hence the reason i run urequest on the second core as it take a relatively long time period

i have used my second core for stuff like polling or directly driving a 7 segment display in software and then [peterhinch](https://github.com/peterhinch) informed me of a way [easier way](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/DRIVERS.md) i can deal with polling and switch bounce and not need my second core (in other words i can use it for something else) i would prefer to not poll my server for data with core 1 like a child saying 'are we there yet' (just need to get data sent from another controller on the network)

**Reply by bixb922** [reply]
OK, now I understand the need

**Reply by GM-Script-Writer-62850** [reply]
i did manage to find [this](https://github.com/DrTom/py-u-async-http-client/tree/master) while looking around just now, maybe i could use that on core 0

i should probably make a bug report since i can't find one... https://github.com/micropython/micropython-lib/issues/700

**Reply by peterhinch** [reply]
That async HTTP client was last updated four years ago and the docs say

> Development of the [uasyncio](https://pypi.org/project/micropython-uasyncio/) library for MicroPython seems to be fragmented and in flux a this time. This project is tested against the [uasyncio](https://pypi.org/project/micropython-uasyncio/) from https://pypi.org/.

That observation is very out of date as is the `uasyncio` version on PyPi. The author claims it works on CPython so it should work on current `uasyncio`. I would definitely avoid the one on PyPi.
